src/text/parsers/top_level.rs

```diff
@@ -1,17 +1,39 @@
 use nom::branch::alt;
 use nom::bytes::streaming::tag;
 use nom::character::streaming::{digit0, one_of};
-use nom::combinator::recognize;
+use nom::combinator::{map, not, recognize};
 
 use crate::text::parse_result::{IonParseResult, OrFatalParseError, UpgradeIResult};
-use nom::sequence::{pair, preceded, tuple};
+use crate::text::parsers::annotations::annotation_delimiter;
+use nom::sequence::{delimited, pair, preceded, tuple};
 use std::str::FromStr;
 
 use crate::text::parsers::comments::whitespace_or_comments;
 
 use crate::text::parsers::value::annotated_value;
 use crate::text::text_value::AnnotatedTextValue;
 
+/// Represents a single item that can appear at the top level of a text stream.
+#[derive(PartialEq, Debug)]
+pub(crate) enum RawTextStreamItem {
+    /// An marker indicating that the stream's version is (major, minor).
+    IonVersionMarker(u32, u32),
+    /// A (possibly annotated) Ion value.
+    AnnotatedTextValue(AnnotatedTextValue),
+}
+
+/// Matches an Ion version marker or a (possibly annotated) value.
+pub(crate) fn stream_item(input: &str) -> IonParseResult<RawTextStreamItem> {
+    alt((
+        map(ion_version_marker, |(major, minor)| {
+            RawTextStreamItem::IonVersionMarker(major, minor)
+        }),
+        map(top_level_value, |value| {
+            RawTextStreamItem::AnnotatedTextValue(value)
+        }),
+    ))(input)
+}
```

Comment:
:world_map: Rather than having the `RawTextReader` check for an `ion_version_marker` and then check for a `top_level_value`, it can now use the `stream_item` parser to consolidate those processes.

Reply:
🔥